modules/aws/ec2-files.go

```diff
@@ -15,85 +16,170 @@ type RemoteFileSpecification struct {
 	RemotePathToFileFilter map[string][]string //A map of the files to fetch, where the keys are directories on the remote host and the values are filters for what files to fetch from the directory. The filters support bash-style wildcards.
 	UseSudo                bool
 	SshUser                string
-	KeyPair                *Ec2Keypair
+	SshOptions             *SshOptions
 	LocalDestinationDir    string //base path where to store downloaded artifacts locally. The final path of each resource will include the ip of the host and the name of the immediate parent folder.
 }
 
+// Specify one of KeyPair, SshAgent, or OverrideSshAgent should be specified
+// Any more (or less) than one will result in error on use
+type SshOptions struct {
+	// Username
+	UserName string
+	// Port
+	Port uint16
+
+	// EC2 Key Pair
+	KeyPair *Ec2Keypair
+	// If true, use default SSH agent on local system (started externally, available at SSH_AUTH_SOCK)
+	SshAgent bool
+	// Override SSH agent on local system
+	OverrideSshAgent *ssh.SshAgent
+
+	// The enabled authentication method
+	// Internally will be one of "keypair", "sshagent", or "overridesshagent"
+	enabledAuthMethod string
+}
+
+func (s *SshOptions) Validate() error {
+	if s.UserName == "" {
+		return fmt.Errorf("SshOptions: UserName field cannot be empty")
+	}
+	if s.Port == 0 {
+		// Default to 22
+		s.Port = 22
+	}
+	if s.KeyPair == nil && s.SshAgent == false && s.OverrideSshAgent == nil {
+		return fmt.Errorf("One of KeyPair, SshAgent or OverrideSshAgent must be set for SshOptions struct")
+	}
+	multipleError := fmt.Errorf("Only one of KeyPair, SshAgent or OverrideSshAgent should be specified in SshOptions struct")
+	if s.KeyPair != nil {
+		if s.SshAgent != false || s.OverrideSshAgent != nil {
+			return multipleError
+		}
+		s.enabledAuthMethod = "keypair"
+		return nil
+	}
+	if s.SshAgent == true {
+		if s.KeyPair != nil || s.OverrideSshAgent != nil {
+			return multipleError
+		}
+		s.enabledAuthMethod = "sshagent"
+		return nil
+	}
+	if s.OverrideSshAgent != nil {
+		if s.KeyPair != nil || s.SshAgent != false {
+			return multipleError
+		}
+		s.enabledAuthMethod = "overridesshagent"
+		return nil
+	}
+	return fmt.Errorf("Unexpected error validating SshOptions struct")
+}
+
+// Attaches the correct authentication method to an ssh.Host struct instance
+func addOptsToSshHost(t *testing.T, sshHost *ssh.Host, sshOpts *SshOptions) {
```

Comment:
Please rename to `addOptsToSshHostE` that returns errors rather than calling `t.Fatal` directly. See the README for details on `E` methods.

Reply:
Also, instead of modifying `sshHost` in place, is it possible to return a new one? Immutability tends to be easier to reason about, especially in an environment with many tests running concurrently.